src/cpp/r/RSexp.cpp

```diff
@@ -1034,36 +1129,209 @@ core::Error getNamespaceExports(SEXP ns,
    return error;
 }
 
-core::Error extractFormalNames(SEXP functionSEXP,
-                               std::vector<std::string>* pNames)
+namespace detail {
+
+bool addSymbolCheckedForMissingness(
+      SEXP nodeSEXP,
+      StringSet* pSymbolsCheckedForMissingness)
 {
+   if (TYPEOF(nodeSEXP) == LANGSXP &&
+       TYPEOF(CAR(nodeSEXP)) == SYMSXP &&
+       CDR(nodeSEXP) != R_NilValue &&
+       TYPEOF(CADR(nodeSEXP)) == SYMSXP &&
+       CDDR(nodeSEXP) == R_NilValue &&
+       strcmp(CHAR(PRINTNAME(CAR(nodeSEXP))), "missing") == 0)
+   {
+      DEBUG("Handling 'missing(" << CHAR(PRINTNAME(CADR(nodeSEXP))) << ")'");
+      pSymbolsCheckedForMissingness->insert(CHAR(PRINTNAME(CADR(nodeSEXP))));
+   }
+   return false;
+}
+
+bool addSymbols(
+      SEXP nodeSEXP,
+      StringSet* pSymbolsUsed)
+{
+   if (TYPEOF(nodeSEXP) == SYMSXP)
+   {
+      DEBUG("Reporting symbol '" << CHAR(PRINTNAME(nodeSEXP)) << "' as used");
+      pSymbolsUsed->insert(CHAR(PRINTNAME(nodeSEXP)));
+   }
+   return false;
+}
+
+void examineSymbolUsage(
+      SEXP nodeSEXP,
+      FunctionSymbolUsage* usage)
+{
+   CallRecurser recurser(nodeSEXP);
+   recurser.add(boost::bind(addSymbols, _1, &(usage->symbolsUsed)));
+   recurser.add(boost::bind(addSymbolCheckedForMissingness, _1,
+                            &(usage->symbolsCheckedForMissingness)));
+   recurser.run();
+}
+
+} // namespace detail
+
+void examineSymbolUsage(
+      SEXP functionSEXP,
+      FunctionInformation* pInfo)
+{
+   if (Rf_isPrimitive(functionSEXP))
+      return;
+   
+   SEXP bodySEXP = BODY_EXPR(functionSEXP);
+   
+   FunctionSymbolUsageCache& cache = functionSymbolUsageCache();
+   FunctionSymbolUsage usage;
+   
+   if (cache.contains(functionSEXP))
+   {
+      usage = cache.get(functionSEXP);
+   }
+   else
+   {
+      detail::examineSymbolUsage(bodySEXP, &usage);
+      cache.put(functionSEXP, usage);
+   }
+   
+   // fill output
+   BOOST_FOREACH(FormalInformation& info, pInfo->formals())
+   {
+      const std::string& name = info.name;
+      info.isUsed = usage.symbolsUsed.contains(name.c_str());
+      
+      bool isInternalFunction = 
+            usage.symbolsUsed.contains(".Internal") ||
+            usage.symbolsUsed.contains(".Primitive");
+      
+      info.missingnessHandled =
+            isInternalFunction ||
+            usage.symbolsCheckedForMissingness.contains(name.c_str());
+   }
+}
+
+class PrimitiveWrappers : boost::noncopyable
+{
+   
+public:
+   
+   SEXP operator[](SEXP primitiveSEXP)
+   {
+      if (contains(primitiveSEXP))
+         return get(primitiveSEXP);
+      
+      r::sexp::Protect protect;
+      SEXP wrapperSEXP;
+      r::exec::RFunction makePrimitiveWrapper(".rs.makePrimitiveWrapper");
+      makePrimitiveWrapper.addParam(primitiveSEXP);
+      Error error = makePrimitiveWrapper.call(&wrapperSEXP, &protect);
+      if (error)
+         LOG_ERROR(error);
+      
+      R_PreserveObject(wrapperSEXP);
```

Comment:
We have the PreservedSEXP class for this but perhaps it's overkill since you will never release these objects (since they are in a static map). Does that sound right?